static/app/components/tabs/tabList.tsx

```diff
@@ -0,0 +1,213 @@
+
+interface TabListProps<T> extends TabListStateProps<T>, AriaTabListProps<T> {
+  className?: string;
+}
+
+/**
+ * Uses IntersectionObserver API to detect overflowing tabs. Returns an array
+ * containing of keys of overflowing tabs.
+ */
+function useOverflowTabs({
+  tabList,
+  tabs,
+}: {
+  tabList: HTMLUListElement | null;
+  tabs: Record<React.Key, HTMLLIElement | null>;
+}) {
+  const [overflowTabs, setOverflowTabs] = useState<React.Key[]>([]);
+
+  useEffect(() => {
+    const options = {
+      root: tabList,
+      // Nagative right margin to account for overflow menu's trigger button
+      rootMargin: `0px -42px 0px ${space(1)}`,
+      threshold: 1,
+    };
+
+    const callback: IntersectionObserverCallback = entries => {
+      entries.forEach(entry => {
+        const {target} = entry;
+        const {key} = (target as HTMLElement).dataset;
+        if (!key) {
+          return;
+        }
+
+        if (!entry.isIntersecting) {
+          setOverflowTabs(prev => prev.concat([key]));
+          return;
+        }
+
+        setOverflowTabs(prev => prev.filter(k => k !== key));
+      });
+    };
+
+    const observer = new IntersectionObserver(callback, options);
+    Object.values(tabs ?? {}).forEach(element => element && observer.observe(element));
+
+    return () => observer.disconnect();
+    // Suppress an exhaustive-deps warning about adding `tabListRef` and
+    // `tabRefs` to the deps array, without knowing that both are ref values
+    // and so don't need to be added.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  return overflowTabs;
+}
+
+/**
+ * To be used as a direct child of the <Tabs /> component. See example usage
+ * in tabs.stories.js
+ */
+export function TabList<T>({className, ...props}: TabListProps<T>) {
+  const tabListRef = useRef<HTMLUListElement>(null);
+  const [disabledKeys, setDisabledKeys] = useState<React.Key[]>([]);
+  const {rootProps, setTabListState} = useContext(TabsContext);
+  const {
+    orientation,
+    disabled,
+    disabledKeys: _disabledKeys,
+    ...otherRootProps
+  } = rootProps;
+  const ariaProps = useMemo(
+    () => ({
+      isDisabled: disabled,
+      disabledKeys,
+      ...otherRootProps,
+      ...props,
+    }),
+    [disabled, disabledKeys, otherRootProps, props]
+  );
+
+  // Load up list state
+  const state = useTabListState(ariaProps);
+  const {tabListProps} = useTabList({orientation, ...ariaProps}, state, tabListRef);
+  useEffect(() => {
+    setTabListState(state);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [state.disabledKeys, state.selectedItem, state.selectedKey, props.children]);
+
+  // Dynamically update `disabledKeys` when the tabs list is loaded/updated
+  useEffect(() => {
+    const newDisabledKeys = [...state.collection]
+      .filter(item => {
+        return item.props.disabled;
+      })
+      .map(item => item.key);
+
+    setDisabledKeys(newDisabledKeys);
+  }, [state.collection]);
```

Comment:
Does this really need to be state? Seems like we can derive this info no?

Reply:
Implemented it this way cause there's a cross-dependency between `disabledKeys` and `state`: `state` uses `disabledKeys` as a prop, but `disabledKeys` needs to iterate over `state.collection` to find disabled items within the collection. A derived `newDisabledKeys` will need to be automatically fed back into `state` as a prop, so it too will need to be a state.